askama/src/filters/humansize.rs

```diff
@@ -43,82 +41,190 @@ impl FastWritable for FilesizeFormatFilter {
         dest: &mut W,
         values: &dyn Values,
     ) -> crate::Result<()> {
-        if self.0 < 1e3 {
+        if self.0 < 1000 {
             (self.0 as u32).write_into(dest, values)?;
             Ok(dest.write_str(" B")?)
-        } else if let Some((prefix, factor)) = SI_PREFIXES
-            .iter()
-            .copied()
-            .find_map(|(prefix_factor, max)| (self.0 < max).then_some(prefix_factor))
-        {
-            // u32 is big enough to hold the number 999_999
-            let scaled = (self.0 * factor) as u32;
-            (scaled / 100).write_into(dest, values)?;
-            format_frac(&mut MaybeUninit::uninit(), prefix, scaled).write_into(dest, values)
         } else {
-            too_big(self.0, dest)
+            // find appropriate unit to format in
+            let unit = SI_PREFIXES
+                .iter()
+                .take_while(|(_, m, _)| *m <= self.0)
+                .last()
+                .unwrap_or(&SI_PREFIXES[SI_PREFIXES.len() - 1]);
+            // divide number up into "full" units, and remainder
+            let integer = self.0 / unit.1;
+            let remainder = self.0 % unit.1;
+
+            // Compute fractional part with desired precision
+            // limit precision to the chosen unit's exponent - a greater precision will just
+            // generated tailing 0s that will be dropped.
+            let precision = self.1.min(unit.2);
+
+            // integer portion (before decimal point)
+            integer.write_into(dest, values)?;
+            if precision > 0 {
+                let mut scale = 10u128.pow(precision as u32);
+                let mut fractional = remainder / (unit.1 / scale);
+                if fractional > 0 {
+                    '.'.write_into(dest, values)?;
+                    for _ in 0..precision {
+                        scale /= 10;
+                        let digit = (b'0' + (fractional / scale) as u8) as char;
+                        digit.write_into(dest, values)?;
+                        fractional %= scale;
+                        if fractional == 0 {
+                            break;
+                        }
+                    }
+                }
+            }
+            ' '.write_into(dest, values)?;
+            unit.0.write_into(dest, values)?;
+            'B'.write_into(dest, values)?;
```

Comment:
```suggestion
            write!(dest, " {}B", unit..0)?;
```

Reply:
Wait really? I'm surprised, generally the I/O calls are slow so we try to do as little of them as possible.

Reply:
The `std` calls are SUPER slow. That's why I use `FastWritable::write_into()` everywhere.
I can do 100s of calls of `FastWritable::write_into()` and will still be faster than one `write!()`.